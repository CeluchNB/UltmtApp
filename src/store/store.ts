import accountReducer from './reducers/features/account/accountReducer'
import { configureStore } from '@reduxjs/toolkit'
import liveGameReducer from './reducers/features/game/liveGameReducer'
import livePointReducer from './reducers/features/point/livePointReducer'
import managedTeamReducer from './reducers/features/team/managedTeamReducer'

const store = configureStore({
    reducer: {
        account: accountReducer,
        managedTeam: managedTeamReducer,
        liveGame: liveGameReducer,
        livePoint: livePointReducer,
    },
})

export type RootState = ReturnType<typeof store.getState>
export default store
