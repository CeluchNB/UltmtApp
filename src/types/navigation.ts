import { DisplayTeam } from './team'
import { DisplayUser } from './user'
import { NativeStackScreenProps } from '@react-navigation/native-stack'
import { RequestType } from './request'

export enum SecureEditField {
    EMAIL,
    PASSWORD,
}

export type TopLevelParamList = {
    Tabs: undefined
    SettingsScreen: undefined
    SecureEditScreen: { title: string; value: string; field: SecureEditField }
}

export type AccountStackParamList = {
    Login: undefined
    Profile: undefined
    CreateAccount: undefined
    CreateTeam: undefined
    ForgotPasswordScreen: undefined
    ManageTeams: undefined
    RequestTeam: undefined
    ManagedTeamDetails: { id: string; place: string; name: string }
    PublicTeamDetails: {
        id: string
        place: string
        name: string
        archive?: boolean
    }
    PublicUserDetails: { user: DisplayUser }
    RequestUser: { type: RequestType }
    ResetPasswordScreen: undefined
    RolloverTeam: undefined
    UserRequestsScreen: undefined
    TeamRequestsScreen: undefined
    JoinByCodeScreen: undefined
}

export type GameStackParamList = {
    GameHome: undefined
    GameSearch: { live: string }
    CreateGame: { teamOne: DisplayTeam }
    SelectOpponent: undefined
}

export type AllScreenProps = NativeStackScreenProps<
    AccountStackParamList & TopLevelParamList & GameStackParamList,
    'Tabs'
>

export type TopLevelProps = NativeStackScreenProps<
    TopLevelParamList,
    'SettingsScreen'
>

export type SecureEditProps = NativeStackScreenProps<
    TopLevelParamList,
    'SecureEditScreen'
>

export type Props = NativeStackScreenProps<AccountStackParamList, 'Login'>

export type CreateTeamProps = NativeStackScreenProps<
    AccountStackParamList,
    'CreateTeam'
>
export type ManagedTeamDetailsProps = NativeStackScreenProps<
    AccountStackParamList,
    'ManagedTeamDetails'
>

export type PublicTeamDetailsProps = NativeStackScreenProps<
    AccountStackParamList,
    'PublicTeamDetails'
>

export type PublicUserDetailsProps = NativeStackScreenProps<
    AccountStackParamList,
    'PublicUserDetails'
>

export type RequestUserProps = NativeStackScreenProps<
    AccountStackParamList,
    'RequestUser'
>

export type GameSearchProps = NativeStackScreenProps<
    GameStackParamList,
    'GameSearch'
>

export type CreateGameProps = NativeStackScreenProps<
    GameStackParamList,
    'CreateGame'
>
