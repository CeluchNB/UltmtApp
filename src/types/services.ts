export type ApiResponse = {
    data: any | undefined
    error: string | undefined
}
