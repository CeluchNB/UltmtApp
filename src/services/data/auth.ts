import * as Constants from '../../utils/constants'
import { ApiError } from '../../types/services'
import { AxiosResponse } from 'axios'
import EncryptedStorage from 'react-native-encrypted-storage'
import jwt_decode from 'jwt-decode'
import { isTokenExpired, throwApiError } from '../../utils/service-utils'
import {
    login as networkLogin,
    logout as networkLogout,
    refreshToken as networkRefreshToken,
} from '../network/auth'

/**
 * Method to login
 * @param username username or email
 * @param password password
 * @returns user token
 * @throws error if backend returns an error
 */
export const login = async (
    username: string,
    password: string,
): Promise<void> => {
    try {
        const response = await networkLogin(username, password)

        const tokens = response.data.tokens
        const { access, refresh } = tokens
        await EncryptedStorage.setItem('access_token', access)
        await EncryptedStorage.setItem('refresh_token', refresh)
    } catch (error: any) {
        return throwApiError(error, Constants.GENERIC_LOGIN_ERROR)
    }
}

/**
 * Logout current user
 * @param token token of current user
 * @throws error if backend returns an error
 */
export const logout = async () => {
    try {
        await withToken(networkLogout)
        await EncryptedStorage.removeItem('access_token')
        await EncryptedStorage.removeItem('refresh_token')
    } catch (error: any) {
        await EncryptedStorage.removeItem('access_token')
        await EncryptedStorage.removeItem('refresh_token')
        throw throwApiError(error, Constants.GENERIC_LOGOUT_ERROR)
    }
}

/**
 * Method to refresh the user's authentication token
 * @returns the new access token
 */
export const refreshToken = async (): Promise<string> => {
    try {
        const token = await EncryptedStorage.getItem('refresh_token')
        if (!token) {
            throw new ApiError(Constants.GENERIC_GET_TOKEN_ERROR)
        }
        const response = await networkRefreshToken(token)
        const { tokens } = response.data
        const { access, refresh } = tokens
        await EncryptedStorage.setItem('access_token', access)
        await EncryptedStorage.setItem('refresh_token', refresh)
        return access
    } catch (error: any) {
        await EncryptedStorage.removeItem('access_token')
        await EncryptedStorage.removeItem('refresh_token')
        throw throwApiError(error, Constants.GENERIC_GET_TOKEN_ERROR)
    }
}

/**
 * Method to make an API call with an authentication token that could be expired.
 * Will make the refresh call and try again if the token is expired.
 * @param networkCall method to call
 * @param args arguments of method
 * @returns expected response of network call
 */
export const withToken = async (
    networkCall: (token: string, ...args: any) => Promise<AxiosResponse>,
    ...args: any
): Promise<AxiosResponse> => {
    try {
        const currentToken =
            (await EncryptedStorage.getItem('access_token')) || ''
        const response = await networkCall(currentToken, ...args)
        return response
    } catch (error: any) {
        const errorJson = error.toJSON()
        if (errorJson.status !== 401) {
            throw error
        }
        try {
            const newToken = await refreshToken()
            return await networkCall(newToken, ...args)
        } catch (error2) {
            throw throwApiError(error2, Constants.GENERIC_GET_TOKEN_ERROR)
        }
    }
}

/**
 * Determine if a user has a valid refresh token
 * @returns true if user logged in
 * @throws error if user is not currently logged in
 */
export const isLoggedIn = async (): Promise<boolean> => {
    try {
        const rToken = await EncryptedStorage.getItem('refresh_token')
        if (!rToken) {
            throw new ApiError(Constants.GENERIC_GET_TOKEN_ERROR)
        }

        const { exp: rExp } = jwt_decode(rToken) as any
        if (isTokenExpired(rExp)) {
            throw new ApiError(Constants.GENERIC_GET_TOKEN_ERROR)
        }

        // TODO: can probably delete this
        const accessToken = await EncryptedStorage.getItem('access_token')
        if (!accessToken) {
            throw new ApiError(Constants.GENERIC_GET_TOKEN_ERROR)
        }
        return true
    } catch (error) {
        throw throwApiError(error, Constants.GENERIC_GET_TOKEN_ERROR)
    }
}

/**
 * Method to determine if a user has a valid access token
 * @returns boolean if user has valid access token
 */
export const hasValidAccessToken = async (): Promise<boolean> => {
    const token = await EncryptedStorage.getItem('access_token')
    if (!token) {
        return false
    }

    const { exp } = jwt_decode(token) as any
    return !isTokenExpired(exp)
}

/**
 * Method to refresh the user's token if necessary. Used for live actions
 * since sockets give no response to determine if a reauth is necessary.
 */
export const refreshTokenIfNecessary = async () => {
    const hasToken = await hasValidAccessToken()
    if (!hasToken) {
        await refreshToken()
    }
}
