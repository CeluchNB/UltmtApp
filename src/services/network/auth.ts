import { ApiResponse } from '../../types/services'
import EncryptedStorage from 'react-native-encrypted-storage'
import { API_KEY, API_URL_V1 } from 'react-native-dotenv'
import axios, { AxiosResponse } from 'axios'

export const login = async (
    username: string,
    password: string,
): Promise<AxiosResponse> => {
    return await axios.post(
        `${API_URL_V1}/auth/login`,
        {
            password,
            email: username,
        },
        {
            headers: {
                'X-API-Key': API_KEY,
            },
        },
    )
}

export const logout = async (token: string): Promise<AxiosResponse> => {
    return await axios.post(
        `${API_URL_V1}/auth/logout`,
        {},
        {
            headers: {
                Authorization: `Bearer ${token}`,
                'X-API-Key': API_KEY,
            },
        },
    )
}

export const refreshToken = async (token: string): Promise<AxiosResponse> => {
    return await axios.post(
        `${API_URL_V1}/auth/refresh`,
        {},
        {
            headers: {
                Authorization: `Bearer ${token}`,
                'X-API-Key': API_KEY,
            },
        },
    )
}

export const getAccessToken = async (): Promise<ApiResponse> => {
    try {
        const token = await EncryptedStorage.getItem('access_token')
        if (token) {
            return { data: token }
        }
    } catch (error) {
        throw error
    }
    throw new Error('No token available')
}
