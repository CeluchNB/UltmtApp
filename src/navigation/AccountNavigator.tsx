import * as React from 'react'
import { AccountStackParamList } from './../types/navigation'
import CreateAccountScreen from './../screens/CreateAccountScreen'
import CreateTeamScreen from './../screens/CreateTeamScreen'
import ForgotPasswordScreen from '../screens/ForgotPasswordScreen'
import LoginScreen from './../screens/LoginScreen'
import ManageTeamDetailsScreen from './../screens/ManageTeamDetailsScreen'
import ManageTeamsScreen from './../screens/ManageTeamsScreen'
import ProfileScreen from './../screens/ProfileScreen'
import PublicTeamScreen from './../screens/PublicTeamScreen'
import PublicUserScreen from './../screens/PublicUserScreen'
import RequestTeamScreen from './../screens/RequestTeamScreen'
import { RequestType } from './../types/request'
import RequestUserScreen from './../screens/RequestUserScreen'
import ResetPasswordScreen from '../screens/ResetPasswordScreen'
import RolloverTeamScreen from './../screens/RolloverTeamScreen'
import { createNativeStackNavigator } from '@react-navigation/native-stack'

const Stack = createNativeStackNavigator<AccountStackParamList>()

const AccountNavigator: React.FC<{}> = () => {
    return (
        <Stack.Navigator
            initialRouteName={'Login'}
            screenOptions={{ headerShown: false }}>
            <Stack.Screen name="Login" component={LoginScreen} />
            <Stack.Screen name="Profile" component={ProfileScreen} />
            <Stack.Screen
                name="CreateAccount"
                component={CreateAccountScreen}
            />
            <Stack.Screen name="CreateTeam" component={CreateTeamScreen} />
            <Stack.Screen name="ManageTeams" component={ManageTeamsScreen} />
            <Stack.Screen name="RequestTeam" component={RequestTeamScreen} />
            <Stack.Screen
                name="ManagedTeamDetails"
                component={ManageTeamDetailsScreen}
                initialParams={{ id: '', place: '', name: '' }}
            />
            <Stack.Screen
                name="PublicTeamDetails"
                component={PublicTeamScreen}
                initialParams={{ id: '', place: '', name: '' }}
            />
            <Stack.Screen
                name="RequestUser"
                component={RequestUserScreen}
                initialParams={{ type: RequestType.PLAYER }}
            />
            <Stack.Screen
                name="RolloverTeam"
                component={RolloverTeamScreen}
                initialParams={{ hasPendingRequests: false }}
            />
            <Stack.Screen
                name="PublicUserDetails"
                component={PublicUserScreen}
                initialParams={{
                    user: {
                        _id: '',
                        firstName: '',
                        lastName: '',
                        username: '',
                    },
                }}
            />
            <Stack.Screen
                name="ForgotPasswordScreen"
                component={ForgotPasswordScreen}
            />
            <Stack.Screen
                name="ResetPasswordScreen"
                component={ResetPasswordScreen}
            />
        </Stack.Navigator>
    )
}

export default AccountNavigator
