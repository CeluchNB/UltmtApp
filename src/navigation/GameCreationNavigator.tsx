import CreateGameScreen from '../screens/games/CreateGameScreen'
import { GameStackParamList } from '../types/navigation'
import React from 'react'
import SelectMyTeamScreen from '../screens/games/SelectMyTeamScreen'
import SelectOpponentScreen from '../screens/games/SelectOpponentScreen'
import { createNativeStackNavigator } from '@react-navigation/native-stack'

const Stack = createNativeStackNavigator<GameStackParamList>()

const GameCreationNavigator: React.FC<{}> = () => {
    return (
        <Stack.Navigator screenOptions={{ headerShown: false }}>
            <Stack.Screen name="SelectMyTeam" component={SelectMyTeamScreen} />
            <Stack.Screen
                name="SelectOpponent"
                component={SelectOpponentScreen}
            />
            <Stack.Screen name="CreateGame" component={CreateGameScreen} />
        </Stack.Navigator>
    )
}

export default GameCreationNavigator
