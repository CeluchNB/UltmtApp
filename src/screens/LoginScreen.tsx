import * as React from 'react'
import * as UserData from '../services/data/user'
import { LoginData } from '../types/reducers'
import PrimaryButton from '../components/atoms/PrimaryButton'
import { Props } from '../types/navigation'
import ScreenTitle from '../components/atoms/ScreenTitle'
import SecondaryButton from '../components/atoms/SecondaryButton'
import UserInput from '../components/atoms/UserInput'
import { useColors } from '../hooks'
import { Controller, useForm } from 'react-hook-form'
import { StyleSheet, Text, View } from 'react-native'
import {
    selectAccount,
    setError,
    setToken,
} from '../store/reducers/features/account/accountReducer'
import { useDispatch, useSelector } from 'react-redux'

const LoginScreen: React.FC<Props> = ({ navigation }: Props) => {
    const hasCheckedLocalToken = React.useRef(false)
    const dispatch = useDispatch()
    const { colors } = useColors()
    const account = useSelector(selectAccount)
    const [loading, setLoading] = React.useState(false)

    const {
        control,
        handleSubmit,
        reset,
        formState: { errors },
    } = useForm({
        defaultValues: {
            username: '',
            password: '',
        } as LoginData,
    })

    const onSubmit = async (data: LoginData) => {
        try {
            setLoading(true)
            reset({
                username: '',
                password: '',
            })
            const { username, password } = data
            const token = await UserData.login(username, password)
            dispatch(setToken(token))
            setLoading(false)
            navigation.navigate('Profile')
        } catch (error: any) {
            setLoading(false)
            dispatch(setError(error.message ?? 'Unable to login'))
        }
    }

    React.useEffect(() => {
        if (!hasCheckedLocalToken.current) {
            UserData.getLocalToken()
                .then(token => {
                    dispatch(setToken(token))
                    navigation.navigate('Profile')
                })
                .catch(error => {
                    dispatch(setError(error.message ?? 'Error fetching token'))
                })
        }
    }, [dispatch, navigation])

    const styles = StyleSheet.create({
        screen: {
            backgroundColor: colors.primary,
            height: '100%',
        },
        title: {
            alignSelf: 'center',
            marginTop: 50,
        },
        input: {
            marginTop: 20,
            width: '75%',
            alignSelf: 'center',
        },
        loginButton: {
            alignSelf: 'center',
            marginTop: 20,
        },
        createButton: {
            alignSelf: 'center',
            marginTop: 5,
        },
        error: {
            color: colors.error,
            width: '75%',
            alignSelf: 'center',
            marginTop: 2,
        },
    })

    return (
        <View style={styles.screen}>
            <ScreenTitle title="Login" style={styles.title} />
            <Controller
                control={control}
                rules={{
                    required: true,
                }}
                render={({ field: { onChange, value } }) => (
                    <UserInput
                        placeholder="Username or Email"
                        onChangeText={onChange}
                        value={value}
                        style={styles.input}
                    />
                )}
                name="username"
            />
            {errors.username && (
                <Text style={styles.error}>This field is required</Text>
            )}
            <Controller
                control={control}
                rules={{ required: true }}
                render={({ field: { onChange, value } }) => (
                    <UserInput
                        placeholder="Password"
                        onChangeText={onChange}
                        value={value}
                        isPassword={true}
                        style={styles.input}
                    />
                )}
                name="password"
            />
            {errors.password && (
                <Text style={styles.error}>This field is required</Text>
            )}
            {account.error && <Text style={styles.error}>{account.error}</Text>}

            <PrimaryButton
                text="Login"
                loading={loading}
                onPress={handleSubmit(onSubmit)}
                style={styles.loginButton}
            />
            <SecondaryButton
                text="Create Account"
                onPress={async () => navigation.navigate('CreateAccount')}
                style={styles.createButton}
            />
        </View>
    )
}

export default LoginScreen
