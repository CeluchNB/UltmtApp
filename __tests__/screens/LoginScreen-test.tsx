import * as AuthData from '../../src/services/data/auth'
import { ApiError } from '../../src/types/services'
import LoginScreen from '../../src/screens/LoginScreen'
import { NavigationContainer } from '@react-navigation/native'
import { Props } from '../../src/types/navigation'
import { Provider } from 'react-redux'
import React from 'react'
import renderer from 'react-test-renderer'
import store from '../../src/store/store'
import { fireEvent, render, waitFor } from '@testing-library/react-native'

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper')

const reset = jest.fn()
const navigate = jest.fn()

const props: Props = {
    navigation: {
        navigate,
        reset,
    } as any,
    route: {} as any,
}

beforeEach(() => {
    reset.mockReset()
    navigate.mockReset()
})

it('test matches snapshot', () => {
    const tree = renderer
        .create(
            <Provider store={store}>
                <LoginScreen {...props} />
            </Provider>,
        )
        .toJSON()
    expect(tree).toMatchSnapshot()
})

it('test successful login', async () => {
    const spy = jest
        .spyOn(AuthData, 'login')
        .mockImplementationOnce(async () => {
            return 'asdf.asdf.asdf'
        })
    const { getByPlaceholderText, getAllByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <LoginScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    fireEvent.changeText(getByPlaceholderText('Username or Email'), 'user')
    fireEvent.changeText(getByPlaceholderText('Password'), 'pass')
    fireEvent.press(getAllByText('Login')[1])

    await waitFor(() => {
        expect(reset).toHaveBeenCalled()
    })

    spy.mockRestore()
})

it('should handle login error', async () => {
    const spy = jest
        .spyOn(AuthData, 'login')
        .mockImplementationOnce(async () => {
            throw new ApiError('error')
        })

    const { getAllByText, getByPlaceholderText, queryByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <LoginScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    fireEvent.changeText(getByPlaceholderText('Username or Email'), 'user')
    fireEvent.changeText(getByPlaceholderText('Password'), 'pass')
    fireEvent.press(getAllByText('Login')[1])

    await waitFor(() => queryByText('error'))
    expect(reset).not.toHaveBeenCalled()
    spy.mockRestore()
})

it('should handle get local token success', async () => {
    const spy = jest
        .spyOn(AuthData, 'getAccessToken')
        .mockImplementationOnce(async () => {
            return '1324.1234.1fgas'
        })

    await waitFor(async () => {
        render(
            <Provider store={store}>
                <NavigationContainer>
                    <LoginScreen {...props} />
                </NavigationContainer>
            </Provider>,
        )
    })

    expect(reset).toHaveBeenCalled()
    spy.mockRestore()
})

it('should handle create profile button press', async () => {
    const { getByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <LoginScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    fireEvent.press(getByText('Create Account'))

    expect(navigate).toHaveBeenCalled()
})

it('should handle forgot password button press', async () => {
    const { getByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <LoginScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    fireEvent.press(getByText('Forgot Password?'))

    expect(navigate).toHaveBeenCalledWith('ForgotPasswordScreen')
})
