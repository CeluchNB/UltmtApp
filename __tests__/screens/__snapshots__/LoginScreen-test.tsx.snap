// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test matches snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
      "height": "100%",
    }
  }
>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "color": "#3183ff",
          "fontSize": 61.64383561643835,
          "fontWeight": "bold",
        },
        Object {
          "alignSelf": "center",
          "marginTop": 50,
        },
      ]
    }
  >
    Login
  </Text>
  <View
    style={
      Object {
        "alignSelf": "center",
        "marginTop": 20,
        "placeholderTextColor": "#303030",
        "width": "75%",
      }
    }
  >
    <View
      pointerEvents="none"
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 6,
          },
          Object {
            "backgroundColor": "#ffffff",
            "borderColor": "#ffd600",
            "borderRadius": 4,
            "borderWidth": 1,
          },
        ]
      }
      testID="text-input-outline"
    />
    <View>
      <View
        style={
          Array [
            Object {
              "paddingBottom": 0,
            },
            Object {
              "minHeight": 64,
              "paddingTop": 8,
            },
          ]
        }
      >
        <TextInput
          editable={true}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Username or Email"
          placeholderTextColor="#303030"
          selectionColor="#3183ff"
          style={
            Array [
              Object {
                "flexGrow": 1,
                "margin": 0,
                "paddingHorizontal": 14,
                "zIndex": 1,
              },
              Object {
                "height": 56,
              },
              Object {
                "paddingBottom": 0,
                "paddingTop": 0,
              },
              Object {
                "color": "#3183ff",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
              false,
              Array [
                Object {},
              ],
            ]
          }
          testID="text-input-outlined"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignSelf": "center",
        "marginTop": 20,
        "placeholderTextColor": "#303030",
        "width": "75%",
      }
    }
  >
    <View
      pointerEvents="none"
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 6,
          },
          Object {
            "backgroundColor": "#ffffff",
            "borderColor": "#ffd600",
            "borderRadius": 4,
            "borderWidth": 1,
          },
        ]
      }
      testID="text-input-outline"
    />
    <View>
      <View
        style={
          Array [
            Object {
              "paddingBottom": 0,
            },
            Object {
              "minHeight": 64,
              "paddingTop": 8,
            },
          ]
        }
      >
        <TextInput
          editable={true}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Password"
          placeholderTextColor="#303030"
          secureTextEntry={true}
          selectionColor="#3183ff"
          style={
            Array [
              Object {
                "flexGrow": 1,
                "margin": 0,
                "paddingHorizontal": 14,
                "zIndex": 1,
              },
              Object {
                "height": 56,
              },
              Object {
                "paddingBottom": 0,
                "paddingTop": 0,
              },
              Object {
                "color": "#3183ff",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
              false,
              Array [
                Object {},
              ],
            ]
          }
          testID="text-input-outlined"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
  </View>
  <View
    collapsable={false}
    nativeID="animatedComponent"
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "#3183ff",
        "borderColor": "transparent",
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 0,
        "elevation": 2,
        "marginTop": 20,
        "minWidth": "auto",
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 0.75,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 1.5,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 4,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "letterSpacing": 1,
                  "marginHorizontal": 16,
                  "marginVertical": 9,
                  "textAlign": "center",
                },
                Object {
                  "marginHorizontal": 8,
                },
                Object {
                  "textTransform": "uppercase",
                },
                Object {
                  "color": "#ffffff",
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                Object {
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                undefined,
              ],
            ]
          }
        >
          Login
        </Text>
      </View>
    </View>
  </View>
  <View
    collapsable={false}
    nativeID="animatedComponent"
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "transparent",
        "borderColor": "#3183ff",
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 0.5,
        "elevation": 0,
        "marginTop": 5,
        "minWidth": 64,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 4,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "letterSpacing": 1,
                  "marginHorizontal": 16,
                  "marginVertical": 9,
                  "textAlign": "center",
                },
                undefined,
                Object {
                  "textTransform": "uppercase",
                },
                Object {
                  "color": "#3183ff",
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                Object {
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                undefined,
              ],
            ]
          }
        >
          Create Account
        </Text>
      </View>
    </View>
  </View>
  <View
    collapsable={false}
    nativeID="animatedComponent"
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "transparent",
        "borderColor": "#3183ff",
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 0.5,
        "elevation": 0,
        "marginTop": 5,
        "minWidth": 64,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 4,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "letterSpacing": 1,
                  "marginHorizontal": 16,
                  "marginVertical": 9,
                  "textAlign": "center",
                },
                undefined,
                Object {
                  "textTransform": "uppercase",
                },
                Object {
                  "color": "#3183ff",
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                Object {
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                undefined,
              ],
            ]
          }
        >
          Forgot Password?
        </Text>
      </View>
    </View>
  </View>
</View>
`;
